core/src/avm1/object.rs

```diff
@@ -338,11 +347,17 @@ impl<'gc> Object<'gc> {
         self.values.contains_key(name)
     }
 
-    pub fn iter_values(&self) -> impl Iterator<Item = (&String, &Value<'gc>)> {
-        self.values.iter().filter_map(|(k, p)| match p {
-            Property::Virtual { .. } => None,
-            Property::Stored { value, .. } => Some((k, value)),
-        })
+    pub fn get_keys(&self) -> Vec<String> {
```

Comment:
Can we have this return `impl Iterator` to avoid the Vec allocation, and let the caller collect if they want to?

Reply:
I see, the the borrow checker really has a fuss when trying this. We want something like this:
```
impl<'a, 'gc: 'a> Object<'gc> {
  pub fn keys<'a>(&'a self) -> Box<dyn 'a + Iterator<Item = String>>
```
because the iterator is only valid for the lifetime of the Object `'a`, and the `'gc` outlives the `Object` (the properties can live longer than the object).

This works totally fine with the `Box`, but trying it with `impl Trait` seems to run into some compiler shortcomings. You get the cryptic error "Hidden type for `impl Trait` captures lifetime that does not appear in bounds":
https://stackoverflow.com/questions/50547766/how-can-i-get-impl-trait-to-use-the-appropriate-lifetime-for-a-mutable-reference 

The dummy trait object method described in the StackOverflow post works:
```
pub fn keys(&'a self) -> impl 'a + Captures<'gc> + Iterator<Item = String>
```
Unless there's a simpler way to do it, that's kind of scary and not sure if it's worth fighting -- unless we're going to need to do this elsewhere too?

Reply:
Gets a little simpler if you just return a reference, and make the caller do the clone if they want:
```
pub fn keys2(&self) -> impl Captures<'gc> + Iterator<Item = &String> {
```
because the lifetime is inferred to be the same for self and the iter item (`'_`), and the caller is responsible for cloning if they need to.

Even further back, the reason why this wasn't an issue before:
```
pub fn iter_values(&self) -> impl Iterator<Item = (&String, &Value<'gc>)> {
```
The `'gc` is baked into the return type again, so the issue doesn't appear. It might be worth having this too so the caller has the option to fetch the value if they want without having to go back through the object (e.g. `locals_into_form_values`)

Reply:
@Herschel one problem we'll have if we return values too (as it did before) is that by reading the values you can modify the object. I think it's better to take a snapshot of the keys before having someone else, if they want to, obtain the values afterwards. Especially considering that everywhere will immediately make a copy of the key to own themselves, and there was only one place that would take the values at the same time.